src/main/java/org/jenkinsci/plugins/github_branch_source/BranchDiscoveryTrait.java

```diff
@@ -0,0 +1,256 @@
+     * Our descriptor.
+     */
+    @Extension
+    @Discovery
+    public static class DescriptorImpl extends SCMSourceTraitDescriptor {
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public String getDisplayName() {
+            return Messages.BranchDiscoveryTrait_displayName();
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean isApplicableToContext(@NonNull Class<? extends SCMSourceContext> contextClass) {
+            return GitHubSCMSourceContext.class.isAssignableFrom(contextClass);
+        }
+
+        /**
+         * Populates the strategy options.
+         *
+         * @return the stategy options.
+         */
+        @NonNull
+        @Restricted(NoExternalUse.class)
+        @SuppressWarnings("unused") // stapler
+        public ListBoxModel doFillStrategyIdItems() {
+            ListBoxModel result = new ListBoxModel();
+            result.add(Messages.BranchDiscoveryTrait_excludePRs(), "1");
+            result.add(Messages.BranchDiscoveryTrait_onlyPRs(), "2");
+            result.add(Messages.BranchDiscoveryTrait_allBranches(), "3");
+            return result;
+        }
+    }
+
+    /**
+     * Trusts branches from the origin repository.
+     */
+    public static class BranchSCMHeadAuthority extends SCMHeadAuthority<SCMSourceRequest, BranchSCMHead, SCMRevision> {
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        protected boolean checkTrusted(@NonNull SCMSourceRequest request, @NonNull BranchSCMHead head) {
+            return true;
+        }
+
+        /**
+         * Out descriptor.
+         */
+        @Extension
+        public static class DescriptorImpl extends SCMHeadAuthorityDescriptor {
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public String getDisplayName() {
+                return Messages.BranchDiscoveryTrait_authorityDisplayName();
+            }
+
+            /**
+             * {@inheritDoc}
+             */
+            @Override
+            public boolean isApplicableToOrigin(@NonNull Class<? extends SCMHeadOrigin> originClass) {
+                return SCMHeadOrigin.Default.class.isAssignableFrom(originClass);
+            }
+        }
+    }
+
+    /**
+     * Filter that excludes branches that are also filed as a pull request.
+     */
+    public static class ExcludeOriginPRBranchesSCMHeadFilter extends SCMHeadFilter {
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean isExcluded(@NonNull SCMSourceRequest request, @NonNull SCMHead head) {
+            if (head instanceof BranchSCMHead && request instanceof GitHubSCMSourceRequest) {
+                for (GHPullRequest pullRequest : ((GitHubSCMSourceRequest) request).getPullRequests()) {
+                    if (pullRequest.getHead().getRef().equals(head.getName())) {
+                        // TODO correct is also a PR test
```

Comment:
I need to re-think whether that TODO is still relevant... I think I was hinting at this trying to detect if one of the other traits would actually find this as a PR rather than just the fact that the branch has been filed as a PR